File: defaults/src/main/java/com/yahoo/vespa/defaults/Defaults.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.defaults;

import java.util.Optional;
import java.util.logging.Logger;

/**
 * The defaults of basic Vespa configuration variables.
 * Use Defaults.getDefaults() to access the defaults of this runtime environment.
 *
 * @author arnej27959
 * @author bratseth
 */
public class Defaults {

    private static final Logger log = Logger.getLogger(Defaults.class.getName());
    private static final Defaults defaults = new Defaults();

    private final String vespaHome;
    private final String vespaUser;
    private final String vespaHost;
    private final String temporaryApplicationStorage;
    private final int vespaWebServicePort;
    private final int vespaPortBase;
    private final int vespaPortConfigServerRpc;
    private final int vespaPortConfigServerHttp;
    private final int vespaPortConfigProxyRpc;

    private Defaults() {
        vespaHome = findVespaHome("/opt/vespa");
        vespaUser = findVespaUser("vespa");
        vespaHost = findVespaHostname("localhost");
        temporaryApplicationStorage = underVespaHome("var/vespa/application");
        vespaWebServicePort = findWebServicePort(8080);
        vespaPortBase = findVespaPortBase(19000);
        vespaPortConfigServerRpc = findConfigServerPort(vespaPortBase + 70);
        vespaPortConfigServerHttp = vespaPortConfigServerRpc + 1;
        vespaPortConfigProxyRpc = findConfigProxyPort(vespaPortBase + 90);
    }

    static private String findVespaHome(String defHome) {
        Optional<String> vespaHomeEnv = Optional.ofNullable(System.getenv("VESPA_HOME"));
        if ( ! vespaHomeEnv.isPresent() || vespaHomeEnv.get().trim().isEmpty()) {
            log.info("VESPA_HOME not set, using " + defHome);
            return defHome;
        }
        String vespaHome = vespaHomeEnv.get().trim();
        if (vespaHome.endsWith("/")) {
            int sz = vespaHome.length() - 1;
            vespaHome = vespaHome.substring(0, sz);
        }
        return vespaHome;
    }

    static private String findVespaHostname(String defHost) {
        Optional<String> vespaHostEnv = Optional.ofNullable(System.getenv("VESPA_HOSTNAME"));
        if (vespaHostEnv.isPresent() && ! vespaHostEnv.get().trim().isEmpty()) {
            return vespaHostEnv.get().trim();
        }
        return defHost;
    }

    static private String findVespaUser(String defUser) {
        Optional<String> vespaUserEnv = Optional.ofNullable(System.getenv("VESPA_USER"));
        if (! vespaUserEnv.isPresent()) {
            log.fine("VESPA_USER not set, using "+defUser);
            return defUser;
        }
        return vespaUserEnv.get().trim();
    }

    static private int findPort(String varName, int defaultPort) {
        Optional<String> port = Optional.ofNullable(System.getenv(varName));
        if ( ! port.isPresent() || port.get().trim().isEmpty()) {
            log.fine("" + varName + " not set, using " + defaultPort);
            return defaultPort;
        }
        try {
            return Integer.parseInt(port.get());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("must be an integer, was '" +
                                               port.get() + "'");
        }
    }

    static private int findVespaPortBase(int defaultPort) {
        return findPort("VESPA_PORT_BASE", defaultPort);
    }
    static private int findConfigServerPort(int defaultPort) {
        return findPort("port_configserver_rpc", defaultPort);
    }
    static private int findConfigProxyPort(int defaultPort) {
        return findPort("port_configproxy_rpc", defaultPort);
    }
    static private int findWebServicePort(int defaultPort) {
        return findPort("VESPA_WEB_SERVICE_PORT", defaultPort);
    }

    /**
     * Get the username to own directories, files and processes
     * @return the vespa user name
     **/
    public String vespaUser() { return vespaUser; }


    /**
     * Compute the host name that identifies myself.
     * Detection of the hostname is now done before starting any Vespa
     * programs and provided in the environment variable VESPA_HOSTNAME;
     * if that variable isn't set a default of "localhost" is always returned.
     * @return the vespa host name
     **/
    public String vespaHostname() { return vespaHost; }

    /**
     * Returns the path where a Vespa application can store arbitrary files on the node. This path 
     * is persistent during the lifetime of this node. The application must be able to recreate
     * required files on its own (e.g. by downloading them from a remote source) if missing.
     *
     * @return the local application storage path
     */
    public String temporaryApplicationStorage() { return temporaryApplicationStorage; }

    /**
     * Returns the path to the root under which Vespa should read and write files.
     * Will not end with a "/".
     *
     * @return the vespa home directory
     */
    public String vespaHome() { return vespaHome; }

    /**
     * Returns an absolute path produced by prepending vespaHome to the argument if it is relative.
     * If the path starts by "/" (absolute) or "./" (explicitly relative - useful for tests),
     * it is returned as-is.
     *
     * @param path the path to prepend vespaHome to unless it is absolute
     * @return the given path string with the root path given from
     *         vespaHome() prepended, unless the given path is absolute, in which
     *         case it is be returned as-is
     */
    public String underVespaHome(String path) {
        if (path.startsWith("/")) return path;
        if (path.startsWith("./")) return path;
        return vespaHome() + "/" + path;
    }

    /**
     * Returns the port number where Vespa web services should be available.
     *
     * @return the vespa webservice port
     */
    public int vespaWebServicePort() { return vespaWebServicePort; }

    /**
     * Returns the base for port numbers where the Vespa services should listen.
     *
     * @return the vespa base number for ports
     */
    public int vespaPortBase() { return vespaPortBase; }

    /** @return port number used by cloud config server (for its RPC protocol) */
    public int vespaConfigServerRpcPort() { return vespaPortConfigServerRpc; }

    /** @return port number used by cloud config server (REST api on HTTP) */
    public int vespaConfigServerHttpPort() { return vespaPortConfigServerHttp; }

    /** @return port number used by config proxy server (RPC protocol) */
    public int vespaConfigProxyRpcPort() { return vespaPortConfigProxyRpc; }

    /** Returns the defaults of this runtime environment */
    public static Defaults getDefaults() { return defaults; }

}

File: client/go/defaults/defaults.go
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
// Author: arnej

package defaults

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/vespa-engine/vespa/client/go/trace"
	"github.com/vespa-engine/vespa/client/go/util"
)

const (
	ENV_VESPA_HOME = util.ENV_VESPA_HOME
	ENV_VESPA_HOST = util.ENV_VESPA_HOSTNAME
	ENV_VESPA_USER = util.ENV_VESPA_USER

	ENV_CONFIGSERVERS     = "VESPA_CONFIGSERVERS"
	ENV_ADDR_CONFIGSERVER = "addr_configserver"

	DEFAULT_VESPA_HOME = "/opt/vespa"
	DEFAULT_VESPA_USER = "vespa"
	DEFAULT_VESPA_HOST = "localhost"

	DEFAULT_VESPA_PORT_BASE = 19000
	ENV_VESPA_PORT_BASE     = "VESPA_PORT_BASE"

	CONFIGSERVER_RPC_PORT_OFFSET = 70
	ENV_CONFIGSERVER_RPC_PORT    = "port_configserver_rpc"

	CONFIGPROXY_RPC_PORT_OFFSET = 90
	ENV_CONFIGPROXY_RPC_PORT    = "port_configproxy_rpc"

	DEFAULT_WEB_SERVICE_PORT = 8080
	ENV_WEB_SERVICE_PORT     = "VESPA_WEB_SERVICE_PORT"
)

// Compute the path prefix where Vespa files will live.
// Note: does not end with "/"
func VespaHome() string {
	if env := os.Getenv(ENV_VESPA_HOME); env != "" {
		return env
	}
	return DEFAULT_VESPA_HOME
}

func UnderVespaHome(p string) string {
	if strings.HasPrefix(p, "/") || strings.HasPrefix(p, "./") {
		return p
	}
	return fmt.Sprintf("%s/%s", VespaHome(), p)
}

// Compute the user name to own directories and run processes.
func VespaUser() string {
	if env := os.Getenv(ENV_VESPA_USER); env != "" {
		return env
	}
	return DEFAULT_VESPA_USER
}

// Compute the host name that identifies myself.
// Detection of the hostname is now done before starting any Vespa
// programs and provided in the environment variable VESPA_HOSTNAME;
// if that variable isn't set a default of "localhost" is always returned.
func VespaHostname() string {
	if env := os.Getenv(ENV_VESPA_HOST); env != "" {
		return env
	}
	return DEFAULT_VESPA_HOST
}

// Compute the port number where the Vespa webservice
// container should be available.
func VespaContainerWebServicePort() int {
	p := getNumFromEnv(ENV_WEB_SERVICE_PORT)
	if p > 0 {
		trace.Debug(ENV_WEB_SERVICE_PORT, p)
		return p
	}
	return DEFAULT_WEB_SERVICE_PORT
}

// Compute the base for port numbers where the Vespa services should listen.
func VespaPortBase() int {
	p := getNumFromEnv(ENV_VESPA_PORT_BASE)
	if p > 0 {
		trace.Debug(ENV_VESPA_PORT_BASE, p)
		return p
	}
	return DEFAULT_VESPA_PORT_BASE
}

// Find the hostnames of configservers that are configured.
func VespaConfigserverHosts() []string {
	parts := splitVespaConfigservers()
	rv := make([]string, len(parts))
	for idx, part := range parts {
		if colon := strings.Index(part, ":"); colon > 0 {
			rv[idx] = part[0:colon]
		} else {
			rv[idx] = part
		}
		trace.Debug("config server host:", rv[idx])
	}
	return rv
}

func findConfigserverHttpPort() int {
	return findConfigserverRpcPort() + 1
}

// Find the RPC addresses to configservers that are configured.
// Returns a list of RPC specs in the format tcp/{hostname}:{portnumber}
func VespaConfigserverRpcAddrs() []string {
	parts := splitVespaConfigservers()
	rv := make([]string, len(parts))
	for idx, part := range parts {
		if colon := strings.Index(part, ":"); colon > 0 {
			rv[idx] = fmt.Sprintf("tcp/%s", part)
		} else {
			rv[idx] = fmt.Sprintf("tcp/%s:%d", part, findConfigserverRpcPort())
		}
		trace.Debug("config server rpc addr:", rv[idx])
	}
	return rv
}

// Find the URLs to the REST api on configservers
// Returns a list of URLS in the format http://{hostname}:{portnumber}/
func VespaConfigserverRestUrls() []string {
	parts := splitVespaConfigservers()
	rv := make([]string, len(parts))
	for idx, hostnm := range parts {
		port := findConfigserverHttpPort()
		if colon := strings.Index(hostnm, ":"); colon > 0 {
			p, err := strconv.Atoi(hostnm[colon+1:])
			if err == nil && p > 0 {
				port = p + 1
			}
			hostnm = hostnm[:colon]
		}
		rv[idx] = fmt.Sprintf("http://%s:%d", hostnm, port)
		trace.Debug("config server rest url:", rv[idx])
	}
	return rv
}

// Find the RPC address to the local config proxy
// Returns one RPC spec in the format tcp/{hostname}:{portnumber}
func VespaConfigProxyRpcAddr() string {
	return fmt.Sprintf("tcp/localhost:%d", findConfigproxyRpcPort())
}

// Get the RPC addresses to all known config sources
// Returns same as vespaConfigProxyRpcAddr + vespaConfigserverRpcAddrs
func VespaConfigSourcesRpcAddrs() []string {
	cs := VespaConfigserverRpcAddrs()
	rv := make([]string, 0, len(cs)+1)
	rv = append(rv, VespaConfigProxyRpcAddr())
	for _, addr := range cs {
		rv = append(rv, addr)
	}
	return rv
}

func splitVespaConfigservers() []string {
	env := os.Getenv(ENV_CONFIGSERVERS)
	if env == "" {
		env = os.Getenv(ENV_ADDR_CONFIGSERVER)
	}
	parts := make([]string, 0, 3)
	for {
		idx := strings.IndexAny(env, " ,")
		if idx < 0 {
			break
		}
		if idx > 0 {
			parts = append(parts, env[:idx])
		}
		env = env[idx+1:]
	}
	if env != "" {
		parts = append(parts, env)
	}
	if len(parts) == 0 {
		parts = append(parts, "localhost")
	}
	return parts
}

func findConfigproxyRpcPort() int {
	p := getNumFromEnv(ENV_CONFIGPROXY_RPC_PORT)
	if p > 0 {
		return p
	}
	return VespaPortBase() + CONFIGPROXY_RPC_PORT_OFFSET
}

func findConfigserverRpcPort() int {
	p := getNumFromEnv(ENV_CONFIGSERVER_RPC_PORT)
	if p > 0 {
		trace.Debug(ENV_CONFIGSERVER_RPC_PORT, p)
		return p
	}
	return VespaPortBase() + CONFIGSERVER_RPC_PORT_OFFSET
}

func getNumFromEnv(vn string) int {
	env := os.Getenv(vn)
	if env != "" {
		p, err := strconv.Atoi(env)
		if err == nil {
			return p
		}
		trace.Debug("env var", vn, "is:", env, "parse error:", err)
	}
	return -1
}
